node/pkg/accountant/audit.go

```diff
@@ -0,0 +1,292 @@
+	for _, pe := range acct.pendingTransfers {
+		if (pe.submitPending) && (time.Since(pe.updTime) > maxSubmitPendingTime) {
+			auditErrors.Inc()
+			acct.logger.Error("acctaudit: transfer has been in the submit pending state for too long", zap.Stringer("lastUpdateTime", pe.updTime))
+		}
+		acct.logger.Debug("acctaudit: will audit pending transfer", zap.String("msgId", pe.msgId), zap.Stringer("lastUpdateTime", pe.updTime))
+		tmpMap[pe.makeAuditKey()] = pe
+	}
+
+	acct.logger.Debug("acctaudit: in AuditPendingTransfers: starting audit", zap.Int("numPending", len(tmpMap)))
+	acct.lastAuditTime = time.Now()
+	go acct.performAudit(tmpMap)
+	acct.logger.Debug("acctaudit: leaving AuditPendingTransfers")
+}
+
+// performAudit audits the temporary map against the smart contract. It is meant to be run in a go routine. It takes a temporary map of all pending transfers
+// and validates that against what is reported by the smart contract. For more details, please see the prologue of this file.
+func (acct *Accountant) performAudit(tmpMap map[string]*pendingEntry) {
+	acct.logger.Debug("acctaudit: entering performAudit")
+	missingObservations, err := acct.queryMissingObservations()
+	if err != nil {
+		acct.logger.Error("acctaudit: unable to perform audit, failed to query missing observations", zap.Error(err))
+		for _, pe := range tmpMap {
+			acct.logger.Error("acctaudit: unsure of status of pending transfer due to query error", zap.String("msgId", pe.msgId))
+		}
+		return
+	}
+
+	if len(missingObservations) != 0 {
+		for _, mo := range missingObservations {
+			key := mo.makeAuditKey()
+			pe, exists := tmpMap[key]
+			if exists {
+				if !pe.submitPending {
+					auditErrors.Inc()
+					acct.logger.Error("acctaudit: contract reported pending observation as missing, resubmitting it", zap.String("msgID", pe.msgId))
+					acct.submitObservation(pe)
+				} else {
+					acct.logger.Info("acctaudit: contract reported pending observation as missing but it is queued up to be submitted, skipping it", zap.String("msgID", pe.msgId))
+				}
+
+				delete(tmpMap, key)
+			} else {
+				acct.handleMissingObservation(mo)
+			}
+		}
+	}
+
+	if len(tmpMap) != 0 {
+		var keys []TransferKey
+		var pendingTransfers []*pendingEntry
+		for _, pe := range tmpMap {
+			keys = append(keys, TransferKey{EmitterChain: uint16(pe.msg.EmitterChain), EmitterAddress: pe.msg.EmitterAddress, Sequence: pe.msg.Sequence})
+			pendingTransfers = append(pendingTransfers, pe)
+		}
+
+		transferDetails, err := acct.queryBatchTransferStatus(keys)
+		if err != nil {
+			acct.logger.Error("acctaudit: unable to finish audit, failed to query for transfer statuses", zap.Error(err))
+			for _, pe := range tmpMap {
+				acct.logger.Error("acctaudit: unsure of status of pending transfer due to query error", zap.String("msgId", pe.msgId))
+			}
+			return
+		}
+
+		for _, pe := range pendingTransfers {
+			item, exists := transferDetails[pe.msgId]
+			if !exists {
+				if !pe.submitPending {
```

Comment:
I was doing the check separately to avoid unnecessary error logging. I reworked it so that submitObservation indicates if it did the submit so I can control logging.